### PR TITLE
Fix Gloo Edge install when failurePolicy is Fail

### PR DESCRIPTION
**Pull request by jenshu**
# Description

The gloo helm charts include gloo custom resources (specifically Gateways) that pass through the validation webhook when they are applied. When webhook failurePolicy is set to Fail, then any issues reaching the validation service would result in an error. This means that it's possible for the install to fail due to attempting to apply custom resources before the validation service is running. 

To allow installs/upgrades to succeed when failurePolicy=Fail, we need to wait until the validation service is ready before applying custom Gateway resources. This is achieved by creating a job on install/upgrade that waits for the gloo deployment to be rolled out, and then applying the Gateway resources in a post-install/post-upgrade hook.

After this change, users will no longer need to do the workaround of [disabling default gateways](https://github.com/solo-io/gloo/pull/6017) when using failurePolicy=Fail.

# Checklist:

- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [ ] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [ ] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/6403

### CODE REVIEW

File: jobs/pkg/kube/secret.go [hunk 1]
```diff
@@ -60,9 +63,17 @@ func GetExistingValidTlsSecret(ctx context.Context, kube kubernetes.Interface, s
 		if now.Before(cert.NotBefore) || now.After(cert.NotAfter) {
 			return nil, nil
 		}
+
+		// check if the cert is valid for this service
+		if certgen.ValidForService(cert.DNSNames, svcName, svcNamespace) {
```
Comment (nfuden) [inline jobs/pkg/kube/secret.go:68]: So as long as any cert is valid for our service then we are happy? 
If thats true then can we short circuit the dns checks for certs after? 

```suggestion
		if !matchesSVC && certgen.ValidForService(cert.DNSNames, svcName, svcNamespace) {
```

Comment (jenshu) [reply jobs/pkg/kube/secret.go]: ah yeah good point, fixed in 44783a4

**Review by nfuden — APPROVED** [review]

File: docs/content/reference/values.txt [hunk 2]
```diff
@@ -334,6 +334,24 @@
 |gateway.certGenJob.cron.enabled|bool|false|enable the cronjob|
 |gateway.certGenJob.cron.schedule|string|* * * * *|Cron job scheduling|
 |gateway.certGenJob.cron.mtlsKubeResourceOverride.NAME|interface||override fields in the gloo-mtls-certgen cronjob.|
+|gateway.rolloutJob.image.tag|string|1.22.9|tag for the container|
```
Comment (artberger) [inline docs/content/reference/values.txt:337]:
```suggestion
|gateway.rolloutJob.image.tag|string|1.22.9| The image tag for the container.|
```

File: docs/content/reference/values.txt [hunk 3]
```diff
@@ -334,6 +334,24 @@
...
+|gateway.rolloutJob.image.repository|string|kubectl|image name (repository) for the container.|
+|gateway.rolloutJob.image.registry|string|bitnami|image prefix/registry e.g. (quay.io/solo-io)|
```
Comment (artberger) [inline docs/content/reference/values.txt:339]:
```suggestion
|gateway.rolloutJob.image.registry|string|bitnami|The image hostname prefix and registry, such as quay.io/solo-io.|
```

File: docs/content/reference/values.txt [hunk 4]
```diff
@@ -334,6 +334,24 @@
...
+|gateway.rolloutJob.image.pullPolicy|string||image pull policy for the container|
```
Comment (artberger) [inline docs/content/reference/values.txt:340]: Is there a default set? Maybe it varies, based on the Kubernetes default?
```suggestion
|gateway.rolloutJob.image.pullPolicy|string||The image pull policy for the container. For default values, see the Kubernetes docs: https://kubernetes.io/docs/concepts/containers/images/#imagepullpolicy-defaulting|
```

File: docs/content/reference/values.txt [hunk 5]
```diff
@@ -334,6 +334,24 @@
...
+|gateway.rolloutJob.image.pullSecret|string||image pull policy for the container |
```
Comment (artberger) [inline docs/content/reference/values.txt:341]: Should this be secret? Is there a default or any other caveat that might be helpful to include?
```suggestion
|gateway.rolloutJob.image.pullSecret|string||The image pull secret to use for the container, in the same namespace as the container pod.|
```

File: docs/content/reference/values.txt [hunk 6]
```diff
@@ -334,6 +334,24 @@
...
+|gateway.rolloutJob.image.extended|bool||if true, deploy an extended version of the container with additional debug tools|
+|gateway.rolloutJob.image.fips|bool||if true, deploy a version of the container (for data-plane containers) built with FIPS-compliant crypto libraries (enterprise only)|
```
Comment (artberger) [inline docs/content/reference/values.txt:343]:
```suggestion
|gateway.rolloutJob.image.fips|bool||If true, deploys a version of the data-plane containers that is built with FIPS-compliant crypto libraries. (Enterprise-only feature.)|
```

File: docs/content/reference/values.txt [hunk 7]
```diff
@@ -334,6 +334,24 @@
...
+|gateway.rolloutJob.floatingUserId|bool||set to true to allow the cluster to dynamically assign a user ID|
```
Comment (artberger) [inline docs/content/reference/values.txt:344]: What is the user ID assigned to?
```suggestion
|gateway.rolloutJob.floatingUserId|bool||If true, allows the cluster to dynamically assign a user ID for the container.|
```

**Review by artberger — COMMENTED** [review]
Left a couple comments from a doc perspective, mostly to use sentence-case consistently. Thanks for adding this fix!

File: docs/content/reference/values.txt (hunk 2 above)
Comment (jenshu) [reply docs/content/reference/values.txt:337]: all of these descriptions on the Jobs were inherited from existing structs. i can fix these but we'll probably need to do another pass to make all the descriptions throughout the file consistent (the source file is `install/helm/gloo/generate/values.go` which generates this values.txt)

File: install/helm/gloo/generate/values.go [hunk 8]
```diff
@@ -261,6 +261,8 @@ type Gateway struct {
 	Validation                    GatewayValidation  `json:"validation,omitempty" desc:"enable Validation Webhook on the Gateway. This will cause requests to modify Gateway-related Custom Resources to be validated by the Gateway."`
 	Deployment                    *GatewayDeployment `json:"deployment,omitempty"`
 	CertGenJob                    *CertGenJob        `json:"certGenJob,omitempty" desc:"generate self-signed certs with this job to be used with the gateway validation webhook. this job will only run if validation is enabled for the gateway"`
+	RolloutJob                    *RolloutJob        `json:"rolloutJob,omitempty" desc:"This job waits for the gloo deployment to be successfully rolled out. It is used to ensure the gateway validation webhook is available before applying Gloo Edge custom resources. This job will only run if gateway validation is enabled and failurePolicy is 'Fail'."`
```
[the remark this answers is not in the document]
Comment (jenshu) [reply install/helm/gloo/generate/values.go:264]: there are several fields that we're checking, so i think it's ok if we keep it vague here with "gateway validation is enabled"

**Comment by jenshu** [comment]
ci is close to passing, so i'll push the docs changes in a separate PR

**Pull request merged**